vmm/src/vstate.rs

```diff
@@ -95,6 +99,52 @@ pub enum Error {
 }
 pub type Result<T> = result::Result<T, Error>;
 
+/// Describes a KVM context that gets attached to the micro vm instance.
+/// It gives access to the functionality of the KVM wrapper as long as every required
+/// KVM capability is present on the host.
+pub struct KvmContext {
+    kvm: Kvm,
+    max_memslots: usize,
+}
+
+impl KvmContext {
+    pub fn new() -> Result<Self> {
+        use kvm_ioctls::Cap::*;
+        let kvm = Kvm::new().expect("Error creating the Kvm object");
+
+        // Check that KVM has the correct version.
+        if kvm.get_api_version() != KVM_API_VERSION as i32 {
+            return Err(Error::KvmApiVersion(kvm.get_api_version()));
+        }
+
+        // A list of KVM capabilities we want to check.
+        #[cfg(target_arch = "x86_64")]
+        let capabilities = vec![Irqchip, Ioeventfd, Irqfd, UserMemory, SetTssAddr];
+
+        #[cfg(target_arch = "aarch64")]
+        let capabilities = vec![Irqchip, Ioeventfd, Irqfd, UserMemory, ArmPsci02];
+
+        // Check that all desired capabilities aer supported.
+        for capability in capabilities.iter() {
+            if !kvm.check_extension(*capability) {
+                return Err(Error::KvmCap(*capability));
+            }
+        }
```

Comment:
I would move these lines into a kvm_validate_caps function or some better name.

Reply:
I gave it a thought and I would prefer not to do it now since long term plan should be to implement a trait that offers functionality that is common to all arch and then implement specific for each arch (x86_64, aarch64) in two different structures. This capability check is architecture specific and IMO the right thing to do move it inside the structures implementing specific functionality